packages/core/src/contrib/Functions.ts

```diff
@@ -1431,6 +1555,195 @@ export const compDict = {
   },
 };
 
+/*
+  Computes the signed distance for a line 
+  float sdSegment( in vec2 p, in vec2 a, in vec2 b )
+  {
+    vec2 pa = p-a, ba = b-a;
+    float h = clamp( dot(pa,ba)/dot(ba,ba), 0.0, 1.0 );
+    return length( pa - ba*h );
+  }
+*/
+const sdLine = (s: Line, p: ad.Num[]): ad.Num => {
+  return sdLineAsNums(s.start.contents, s.end.contents, p);
+};
+
+const sdLineAsNums = (a: ad.Num[], b: ad.Num[], p: ad.Num[]): ad.Num => {
+  const pa = ops.vsub(p, a);
+  const ba = ops.vsub(b, a);
+  const h = clamp([0, 1], div(ops.vdot(pa, ba), ops.vdot(ba, ba)));
+  return ops.vnorm(ops.vsub(pa, ops.vmul(h, ba)));
+};
+
+const sdPolyline = (s: Polyline, p: ad.Num[]): ad.Num => {
+  const dists: ad.Num[] = [];
+  for (let i = 0; i < s.points.contents.length - 1; i++) {
+    const start = s.points.contents[i];
+    const end = s.points.contents[i + 1];
+    dists[i] = sdLineAsNums(start, end, p);
+  }
+  return minN(dists);
+};
+
+export const sdEllipse = (s: Ellipse, p: ad.Num[]): ad.Num => {
+  return sdEllipseAsNums(s.rx.contents, s.ry.contents, s.center.contents, p);
+};
+
+/*
+  float msign(in float x) { return (x<0.0)?-1.0:1.0; }
+  TODO: merge with Jiri's signOf function
+  */
+const msign = (x: ad.Num): ad.Num => {
+  return ifCond(lt(x, 0), -1, 1);
+};
+
+/*
+  p = abs( p ); 
+  if( p.x>p.y ){ p=p.yx; ab=ab.yx; }
+	
+	float l = ab.y*ab.y - ab.x*ab.x;
+	
+  float m = ab.x*p.x/l; 
+	float n = ab.y*p.y/l; 
+	float m2 = m*m;
+	float n2 = n*n;
+	
+    float c = (m2+n2-1.0)/3.0; 
+	float c3 = c*c*c;
+
+    float d = c3 + m2*n2;
+    float q = d  + m2*n2;
+    float g = m  + m *n2;
+
+    float co;
+
+    if( d<0.0 )
+    {
+        float h = acos(q/c3)/3.0;
+        float s = cos(h) + 2.0;
+        float t = sin(h) * sqrt(3.0);
+        float rx = sqrt( m2-c*(s+t) );
+        float ry = sqrt( m2-c*(s-t) );
+        co = ry + sign(l)*rx + abs(g)/(rx*ry);
+    }
+    else
+    {
+        float h = 2.0*m*n*sqrt(d);
+        float s = msign(q+h)*pow( abs(q+h), 1.0/3.0 );
+        float t = msign(q-h)*pow( abs(q-h), 1.0/3.0 );
+        float rx = -(s+t) - c*4.0 + 2.0*m2;
+        float ry =  (s-t)*sqrt(3.0);
+        float rm = sqrt( rx*rx + ry*ry );
+        co = ry/sqrt(rm-rx) + 2.0*g/rm;
+    }
+    co = (co-m)/2.0;
+
+    float si = sqrt( max(1.0-co*co,0.0) );
+ 
+    vec2 r = ab * vec2(co,si);
+	
+    return length(r-p) * msign(p.y-r.y);
+}
+*/
```

Comment:
Where exactly does this reference code come from? It does not match the `sdEllipse` code on this page: https://iquilezles.org/articles/distfunctions2d/

Reply:
It is from the actual shadertoy code that is linked from the iquielezes blog post (https://www.shadertoy.com/view/4sS3zz). 
-Josh